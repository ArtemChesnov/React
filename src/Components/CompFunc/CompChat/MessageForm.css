.mesageFromUser-form {
    margin: 50px 0 20px 15px;
    display: flex;
    justify-content: space-evenly;
}

.message-board {
    background-color: rgba(51, 230, 215, 0.7);
    width: 60vh;
    height: 90vh;
    border: 1px;
    border-radius: 30px;
}

.message-chat {
    background-color: rgba(179, 216, 213, 0.9);
    border-radius: 15px;
    margin: 15px 10px 15px 10px;
    height: 80%;
    overflow: scroll;
    overflow-x: hidden;
}

.message-chat::-webkit-scrollbar {
    width: 16px;
}

.message-chat::-webkit-scrollbar-track {
    background-color: rgba(179, 216, 213, 0.9);
    border-radius: 100px;
}

.message-chat::-webkit-scrollbar-thumb {
    background-color: rgba(3, 87, 78, 0.9);
    border-radius: 100px;
}

.message-list {
    margin: 0;
    padding: 0;
    display: flex;
    flex-direction: column;
    list-style-type: none;
    text-align: left;
}

.message-item {
    margin: 15px 10px 0 15px;
    overflow: hidden;
}

.message-item:first-child {
    margin-top: 45px;
}

.message-autor {
    font-size: 24px;
    text-transform: capitalize;
    color: rgb(0, 0, 0);
    margin: 0 25px 5px 0;
    text-align: right;
}

.message-text {
    word-break: break-all;
    border: 2px solid rgba(51, 230, 215, 0.7);
    background-color: rgba(68, 206, 194, 0.3);
    border-radius: 5px;
    padding: 10px 10px 25px 10px;
    font-size: calc(5px + 2vmin);
    color: rgb(37, 37, 37);

    margin: 0 0 0 10px;
    position: relative;
}

.message-time {
    position: absolute;
    right: 5px;
    bottom: 5px;
}

.message-item:nth-child(odd) .message-autor {
    margin: 0 0 5px 25px;
    text-align: left;
}

.message-item:nth-child(odd) .message-text {
    border-color: rgba(51, 230, 215, 0.7);
    background-color: rgba(68, 206, 194, 0.6);
    margin: 0 10px 0 0;
}

.message-form {
    display: flex;
    flex-direction: column;
    margin: 50px;
}