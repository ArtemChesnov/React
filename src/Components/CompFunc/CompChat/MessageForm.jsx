import React, { useState, useEffect, useCallback } from 'react';
import { AutorInput } from './AutorInput';
import { MessageButton } from './MessageButton';
import { MessageInput } from './MessageInput';
import { MessageList } from './MessageList';

export const MessageForm = () => {
  const [autorValue, setAutorValue] = useState('');
  const [messageValue, setMessageValue] = useState('');
  const [message, setMessage] = useState([]);

  const clickSubmit = useCallback(() => {
    setMessage([
      ...message,
      {
        autor: autorValue,
        value: messageValue,
        now: new Date().toLocaleTimeString().slice(0, -3),
      },
    ]);
    setAutorValue('');
    setMessageValue('');
  }, [message, autorValue, messageValue]);

  useEffect(() => {
    if (message.length > 0 && message[message.length - 1].autor !== 'Душнила') {
      const timeout = setTimeout(() => {
        setMessage([
          ...message,
          {
            autor: 'Душнила',
            value: `Привет ${message[message.length - 1].autor}, ты скучный.`,
            now: new Date().toLocaleTimeString().slice(0, -3),
          },
        ]);
      }, 2000);
      return () => {
        clearTimeout(timeout);
      };
    }
  }, [message]);

  const autorChange = (event) => {
    setAutorValue(event.target.value);
  };

  const messageChange = (event) => {
    setMessageValue(event.target.value);
  };

  return (
    <form>
      <div className="mesageFromUser-form">
        <MessageList message={message} />
        <div className="message-form">
          <AutorInput autorValue={autorValue} changeName={autorChange} />
          <MessageInput
            changeMessage={messageChange}
            messageValue={messageValue}
          />
          <MessageButton
            disabled={!messageValue || !autorValue}
            click={clickSubmit}
          />
        </div>
      </div>
    </form>
  );
};
