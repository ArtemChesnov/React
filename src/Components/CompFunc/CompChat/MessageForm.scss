.mesageFromUser-form {
    margin: 50px 0 20px 15px;
    display: flex;
    justify-content: space-evenly;
}

.message-form {
    display: flex;
    flex-direction: column;
    margin: 50px;
}