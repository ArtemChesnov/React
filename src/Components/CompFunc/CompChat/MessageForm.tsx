import React, { FC, useState, memo } from 'react';
import { useParams } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { addMessageWithReply } from '../../../store/chats/slice';
import { ChatsState } from 'src/store/chats/reducer';
import { AddMessage } from 'src/store/chats/types';
import { ThunkDispatch } from 'redux-thunk';

import { AuthorInput } from './AuthorInput/AuthorInput';
import { MessageButton } from './MessageButton/MessageButton';
import { MessageInput } from './MessageInput/MessageInput';

import style from './MessageForm.module.scss';

export const MessageForm: FC = memo(() => {
  const [authorValue, setAuthorValue] = useState('');
  const [messageValue, setMessageValue] = useState('');

  const dispatch =
    useDispatch<ThunkDispatch<ChatsState, void, ReturnType<AddMessage>>>();
  const { chatId } = useParams();

  const handleSubmitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (chatId && messageValue) {
      dispatch(
        addMessageWithReply({
          chatId,
          message: {
            author: authorValue,
            value: messageValue,
            now: new Date().toLocaleTimeString().slice(0, -3),
          },
        })
      );
    }

    setAuthorValue('');
    setMessageValue('');
  };

  return (
    <form onSubmit={handleSubmitForm} className={style.form} data-testid="form">
      <AuthorInput authorValue={authorValue} setAuthorValue={setAuthorValue} />
      <MessageInput
        setMessageValue={setMessageValue}
        messageValue={messageValue}
      />
      <MessageButton disabled={!messageValue || !authorValue} />
    </form>
  );
});
