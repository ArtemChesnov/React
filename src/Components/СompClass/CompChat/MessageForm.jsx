import React, { Component } from 'react';
import { AutorInput } from './AutorInput';
import { MessageButton } from './MessageButton';
import { MessageInput } from './MessageInput';

export class MessageForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      autorValue: '',
      messageValue: '',
      message: [],
    };
  }

  clickSubmit = () => {
    this.setState({
      message: [
        ...this.state.message,
        {
          autor: this.state.autorValue,
          message: this.state.messageValue,
          now: new Date().toLocaleTimeString().slice(0, -3),
        },
      ],
    });
    this.setState({ messageValue: '' });
    this.setState({ autorValue: '' });
  };

  bot = () => {
    this.setState({
      message: [
        ...this.state.message,
        {
          autor: 'Душнила',
          message: `Привет ${
            this.state.message[this.state.message.length - 1].autor
          }, ты скучный.`,
          now: new Date().toLocaleTimeString().slice(0, -3),
        },
      ],
    });
  };

  componentDidUpdate() {
    if (
      this.state.message.length > 0 &&
      this.state.message[this.state.message.length - 1].autor !== 'Душнила'
    ) {
      this.interval = setTimeout(this.bot, 1500);
    }
  }

  componentWillUnmount() {
    clearTimeout(this.interval);
  }

  autorChange = (event) => {
    this.setState({ autorValue: event.target.value });
  };

  messageChange = (event) => {
    this.setState({ messageValue: event.target.value });
  };

  render() {
    return (
      <div className="mesageFromUser-form">
        <div className="message-board">
          <h1 className="message-title">Чат с Душнилой</h1>
          <div className="message-chat">
            <ul className="message-list">
              {this.state.message.map((message, idx) => (
                <li key={idx} className="message-item">
                  <h1 className="message-autor">{message.autor}</h1>
                  <p className="message-text">
                    {message.message}
                    <span className="message-time">{message.now}</span>
                  </p>
                </li>
              ))}
            </ul>
          </div>
        </div>
        <div className="message-form">
          <AutorInput
            autorValue={this.state.autorValue}
            changeName={this.autorChange}
          />
          <MessageInput
            changeMessage={this.messageChange}
            messageValue={this.state.messageValue}
          />
          <MessageButton
            disabled={!this.state.messageValue || !this.state.autorValue}
            click={this.clickSubmit}
          />
        </div>
      </div>
    );
  }
}
