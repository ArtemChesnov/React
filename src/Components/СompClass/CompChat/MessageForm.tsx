import React, { Component } from 'react';
import { AutorInput } from './AutorInput';
import { MessageButton } from './MessageButton';
import { MessageInput } from './MessageInput';

interface FormProps {
  addMessage: (autorValue: string, messageValue: string) => void;
}
interface FormState {
  autorValue: string;
  messageValue: string;
}

export class MessageForm extends Component<FormProps, FormState> {
  constructor(props: FormProps) {
    super(props);
    this.state = {
      autorValue: '',
      messageValue: '',
    };
  }
  clickSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.props.addMessage(this.state.messageValue, this.state.autorValue);
    this.setState({ messageValue: '' });
    this.setState({ autorValue: '' });
  };

  messageChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    this.setState({ messageValue: event.target.value });
  };

  autorChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ autorValue: event.target.value });
  };

  render() {
    return (
      <form onSubmit={this.clickSubmit}>
        <div className="mesageFromUser-form">
          <div className="message-form">
            <AutorInput
              autorValue={this.state.autorValue}
              autorChange={this.autorChange}
            />
            <MessageInput
              messageChange={this.messageChange}
              messageValue={this.state.messageValue}
            />
            <MessageButton
              disabled={!this.state.messageValue || !this.state.autorValue}
            />
          </div>
        </div>
      </form>
    );
  }
}
