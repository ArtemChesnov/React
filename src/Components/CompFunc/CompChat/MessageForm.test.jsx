import React from 'react';
import { MessageForm } from './MessageForm';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
// import { userEvent } from '@testing-library/user-event';
// import { waitFor } from '@storybook/testing-library';

describe('MessageForm', () => {
  it('render component', () => {
    render(<MessageForm />);
  });

  it('render with snapshot', () => {
    const { asFragment } = render(<MessageForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('render component with text', () => {
    render(<MessageForm />);
    expect(
      screen.getByPlaceholderText('Введите ваше имя...')
    ).toBeInTheDocument();
  });
});
