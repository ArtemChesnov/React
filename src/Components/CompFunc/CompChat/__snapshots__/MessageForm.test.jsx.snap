// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm render with snapshot 1`] = `
<DocumentFragment>
  <form
    data-testid="form"
  >
    <div
      class="mesageFromUser-form"
    >
      <div
        class="message-form"
      >
        <input
          class="autor-input"
          placeholder="Введите ваше имя..."
          required=""
          type="text"
          value=""
        />
        <textarea
          class="message-input"
          placeholder="Напишите сообщение..."
          required=""
          type="text"
        />
        <button
          class="message-button"
          disabled=""
          id="button"
          type="submit"
        >
          Отправить
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
